File: Tutorial4-3ComponentBuilders/src/main/java/com/example/tutorial4_3componentbuilders/di/AppComponent.java
package com.example.tutorial4_3componentbuilders.di;

import com.example.tutorial4_3componentbuilders.MainActivity;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(modules = {AppModule.class})
public interface AppComponent {
    void inject(MainActivity mainActivity);

    // DaggerAppComponent.build() returns this Builder interface

/*    @Component.Builder
    interface Builder {
        AppComponent build();
        Builder appModule(AppModule appModule);
    }*/

}

File: Tutorial4-3ComponentBuilders/src/main/java/com/example/tutorial4_3componentbuilders/di/AppComponent.kt
package com.example.tutorial4_3componentbuilders.di

import com.example.tutorial4_3componentbuilders.MainActivity
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [AppModule::class])
interface AppComponent {

    fun inject(mainActivity: MainActivity)

    // DaggerAppComponent.build() returns this Builder interface
    @Component.Builder
    interface Builder {

        fun build(): AppComponent

        fun appModule(appModule: AppModule): Builder
    }

}